src/block_data.rs

```diff
@@ -0,0 +1,320 @@
+            .map_err(BlockConversionError::WrongHeader)?;
+        
+        // Construct Transactions and Receipts from dynamic datum indexes
+        let (txns_bs, receipts_bs) = DatumIndexV2::transactions_and_receipts(hotstuff_data)
+            .ok_or(BlockConversionError::IncorrectNumberOfTxsAndReceipts)?;
+
+        // Deserialize transactions (and validate the signatures)
+        let mut transactions = Vec::with_capacity(txns_bs.len());
+        for txn_bs in txns_bs {
+            let txn: TransactionV2 = Deserializable::deserialize(txn_bs)
+                .map_err(|_| BlockConversionError::Transaction)?;
+            // Check transactions signature and return error immediately if there is an invalid transaction.
+            if verify_transaction_signatures {
+                txn.is_cryptographically_correct()
+                    .map_err(|_| BlockConversionError::InvalidTransactionSignature)?;
+            }
+            transactions.push(txn);
+        }
+
+        // Deserialize receipts
+        let mut receipts: Vec<ReceiptV2> = Vec::with_capacity(receipts_bs.len());
+        for receipt_bs in receipts_bs {
+            receipts.push(
+                Deserializable::deserialize(receipt_bs)
+                    .map_err(|_| BlockConversionError::Receipt)?,
+            )
+        }
+
+        Ok(BlockDataV2{
+            header,
+            transactions,
+            receipts,
+        })
+    }
+}
+
+impl From<BlockDataV2> for Data {
+    fn from(value: BlockDataV2) -> Data {
+        let header = Data::from(value.header);
+
+        let transactions: Data = value
+            .transactions
+            .into_iter()
+            .map(|tx| Serializable::serialize(&tx))
+            .collect();
+
+        let receipts: Data = value
+            .receipts
+            .into_iter()
+            .map(|recp| Serializable::serialize(&recp))
+            .collect();
+
+        [header, transactions, receipts].concat()
+    }
+}
+
+impl<'a> TryFrom<&'a Data> for BlockDataV2 {
+    type Error = BlockConversionError;
+    fn try_from(data: &'a Data) -> std::result::Result<Self, Self::Error> {
+        BlockDataV2::from_data(data, true)
+    }
+}
+
+
+/// A data structure that encapsulates the data of block header from [hotstuff_rs::types::Data].
+#[derive(Clone, BorshSerialize, BorshDeserialize)]
+pub struct BlockHeaderDataV2 {
+    pub chain_id: hotstuff_rs::types::ChainID,
+    pub proposer: PublicAddress,
+    pub timestamp: u32,
+    pub base_fee_per_gas: u64,
+    pub gas_used: u64,
+    pub transactions_hash: Sha256Hash,
+    pub receipts_hash: Sha256Hash,
+    pub state_hash: Sha256Hash,
+    pub logs_bloom: BloomFilter,
+}
+impl Serializable for BlockHeaderDataV2 { }
+
+impl BlockHeaderDataV2 {
+    pub fn hash(&self) -> hotstuff_rs::types::CryptoHash {
+        sha256(self.try_to_vec().unwrap())
+    }
+}
+
+impl TryFrom<&Data> for BlockHeaderDataV2 {
+    type Error = BlockHeaderConversionError;
```

Comment:
Likewise, I think `BlockHeaderDataFromHotStuffDataError` is a better name for this type.